File: src/main/java/club/geek66/querydsl/OperatorUtils.java
package club.geek66.querydsl;

import com.querydsl.core.types.Ops;
import io.vavr.collection.HashMap;
import io.vavr.control.Either;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import lombok.experimental.UtilityClass;


/**
 * @author: orange
 * @date: 2020/12/3
 * @time: 下午4:11
 * @copyright: Copyright 2020 by orange
 */
@UtilityClass
public class OperatorUtils {

	@Data
	@RequiredArgsConstructor
	static class ParseOpsError {

		private final String msg;

	}

	public Either<ParseOpsError, Ops> parseOps(String operator) {
		return aliasMap.get(operator).toEither(new ParseOpsError("Nonsupport operator " + operator));
	}

	private final HashMap<String, Ops> aliasMap = HashMap.of(
			">", Ops.GT, "GT", Ops.GT,
			"<", Ops.LT, "LT", Ops.LT,
			">=", Ops.GOE, "GOE", Ops.GOE,
			"<=", Ops.LOE, "LOE", Ops.LOE,
			"==", Ops.EQ, "EQ", Ops.EQ
	).put("!=", Ops.NE).put("NE", Ops.NE);

}

File: src/main/kotlin/club/geek66/querydsl/OperatorUtils.kt
package club.geek66.querydsl

import arrow.core.Either
import arrow.core.toOption
import com.querydsl.core.types.Ops

/**
 * @author: orange
 * @date: 2020/12/3
 * @time: 下午4:11
 * @copyright: Copyright 2020 by orange
 */
object OperatorUtils {

	class ParseOpsError(
		val msg: String
	)

	fun parseOps(operator: String): Either<ParseOpsError, Ops> =
		aliasMap[operator]
			.toOption()
			.toEither { ParseOpsError("Nonsupport operator $operator") }

	private val aliasMap = mapOf(
		">" to Ops.GT, "GT" to Ops.GT,
		"<" to Ops.LT, "LT" to Ops.LT,
		">=" to Ops.GOE, "GOE" to Ops.GOE,
		"<=" to Ops.LOE, "LOE" to Ops.LOE,
		"==" to Ops.EQ, "EQ" to Ops.EQ,
		"!=" to Ops.NE, "NE" to Ops.NE
	)
}